File: 37-sudoku-solver/sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        rows = [0] * 9
        cols = [0] * 9
        boxes = [0] * 9
        
        def box_index(r, c):
            return (r // 3) * 3 + c // 3
        
        # Initialize bitmasks
        for r in range(9):
            for c in range(9):
                if board[r][c] != '.':
                    digit = int(board[r][c]) - 1
                    mask = 1 << digit
                    rows[r] |= mask
                    cols[c] |= mask
                    boxes[box_index(r, c)] |= mask
        
        def backtrack(row, col):
            if row == 9:
                return True
            
            next_row, next_col = (row + 1, 0) if col == 8 else (row, col + 1)
            
            if board[row][col] != '.':
                return backtrack(next_row, next_col)
            
            b = box_index(row, col)
            used = rows[row] | cols[col] | boxes[b]
            
            for digit in range(9):
                mask = 1 << digit
                if used & mask:
                    continue
                
                # Place digit
                board[row][col] = str(digit + 1)
                rows[row] |= mask
                cols[col] |= mask
                boxes[b] |= mask
                
                if backtrack(next_row, next_col):
                    return True
                
                # Backtrack
                board[row][col] = '.'
                rows[row] &= ~mask
                cols[col] &= ~mask
                boxes[b] &= ~mask
            
            return False
        
        backtrack(0, 0)

File: 37-sudoku-solver/sudoku-solver.cpp
class Solution {
public:
    int rows[9] = {0}, cols[9] = {0}, boxes[9] = {0};
    
    void solveSudoku(vector<vector<char>>& board) {
        // Initialize bitmasks: bit k set means digit (k+1) is used
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    int digit = board[i][j] - '1'; // 0-indexed digit
                    int mask = 1 << digit;
                    rows[i] |= mask;
                    cols[j] |= mask;
                    boxes[(i/3)*3 + j/3] |= mask;
                }
            }
        }
        backtrack(board, 0, 0);
    }
    
    bool backtrack(vector<vector<char>>& board, int row, int col) {
        if (row == 9) return true;
        
        int nextRow = (col == 8) ? row + 1 : row;
        int nextCol = (col == 8) ? 0 : col + 1;
        
        if (board[row][col] != '.') {
            return backtrack(board, nextRow, nextCol);
        }
        
        int b = (row/3)*3 + col/3;
        int used = rows[row] | cols[col] | boxes[b];
        
        for (int digit = 0; digit < 9; digit++) {
            int mask = 1 << digit;
            if (used & mask) continue; // digit already used
            
            // Place digit
            board[row][col] = '1' + digit;
            rows[row] |= mask;
            cols[col] |= mask;
            boxes[b] |= mask;
            
            if (backtrack(board, nextRow, nextCol)) return true;
            
            // Backtrack
            board[row][col] = '.';
            rows[row] &= ~mask;
            cols[col] &= ~mask;
            boxes[b] &= ~mask;
        }
        
        return false;
    }
};